### DISCUSSION TITLE
Responses "Undefined" For some Survey-Likert Items

### DISCUSSION THREAD

**Discussion by csducate** [jspsych/jsPsych#842] [Q&A]
Hello,

I am currently coding a survey with some Likert-like items. However, some of my items return data with improbable response times and "undefined" responses. I cannot figure out what is going wrong because the items display properly & do not throw errors when I test them in the MTurk sandbox. 

My code for one set of items looks like this:

```javascript
var vignetteArray = [jsPsych.randomization.sampleWithoutReplacement([
          {
            vignette1: 'BLAH BLAH BLAH',
            data: {condition: 'neutral'}
          },
          {
            vignette2: 'MORE BLAH'
            data: {condition: 'coercive'}
          }
        ], 1)[0]
      ];

// Variable with the likert scale used
var scale1 = ["Extremely Unlikely", "Somewhat Unlikely", "Neither Likely nor Unlikely", "Somewhat Likely", "Extremely Likely"];

var scaleWidth = 800;

// Behavioral about response to vignette
var likertRestaurant1 = {
    type: 'survey-likert',
    questions: [
        {
          prompt: "Complain about the food, drinks, or service to a manager.",
          labels: scale1,
          data: {QN: 'Q4.1'}
        },
        {
          prompt: "Demand that items be taken off your bill.",
          labels: scale1,
          data: {QN: 'Q4.2'}
        },
        {
          prompt: "Return to this restaurant at a later date.",
          labels: scale1,
          data: {QN: 'Q4.3'}
        },
        {
          prompt: "Recommend this restaurant to a friend or family member.",
          labels: scale1,
          data: {QN: 'Q4.4'}
        },
        {
          prompt: "Tell others not to go to this restaurant.",
          labels: scale1,
          data: {QN: 'Q4.5'}
        }
    ],
    preamble: jsPsych.timelineVariable('restaurantVig'),
    scale_width: scaleWidth,
    on_finish: function(data){
          console.log(JSON.stringify(data));
        }
};
```

So the code successfully pulls one of two vignettes as the preamble for 5 Likert-like questions. However, when I look at the data saved from this item set, it looks like this:

```javascript
	{
		"rt": 12672.90000000503,
		"responses": "{\"undefined\":3}",
		"question_order": "[0,1,2,3,4]",
		"trial_type": "survey-likert",
		"trial_index": 90,
		"time_elapsed": 112221,
		"internal_node_id": "0.0-6.0-0.0"
	}
```

Does anyone know what is happening? It seems to be an issue with `survey-likert` **blocks** of questions, but I don't know why. Why would the code display the items but not save the data?

**Comment by jodeleeuw** [comment]
Hi @csducate,

I'm having trouble replicating this issue. Here's what the data look like for me:

```js
{
		"rt": 7757.350000028964,
		"responses": "{\"Q0\":1,\"Q1\":1,\"Q2\":3,\"Q3\":4,\"Q4\":2}",
		"question_order": "[0,1,2,3,4]",
		"trial_type": "survey-likert",
		"trial_index": 0,
		"time_elapsed": 7760,
		"internal_node_id": "0.0-0.0"
	}
```

I tested this with jsPsych 6.1.0 and Chrome. I don't think the browser would make a difference here, but I've been surprised before.

**Reply by csducate** [reply]
Hi Josh,

That's...weird.

I was using jsPsych 6.0.5 and Chrome. I updated to jsPsych 6.1.0 (so I should have clean code), but I can only make the problem go away if I explicitly set the name parameter. If I omit the name parameter, that question is "undefined," not auto-named as Q0, Q1, etc.

This ends up working for my purposes, as I did want to name my questions, but I have no idea why it is happening for me. As far as I'm aware, it hasn't always happened with this code--it's a relatively new problem. I can replicate the problem in Chrome Incognito Mode and in Firefox. I'm not sure if it's an issue with the name parameter itself because I had the same issue with 6.0.5.

I guess I'll just use the name parameter for all survey items for now.

Caitlin

**Reply by jodeleeuw** [reply]
Hi Caitlin,

When you updated to jsPsych 6.1.0 did you update the plugin file too? I'm guessing yes, but I'm just trying to track down if there is a bug here that we need to be finding.

Thanks!
Josh

**Reply by csducate** [reply]
Hi Josh,

Yes, I updated all of the plugins I use. They have a creation date of July 15, 2019, which seems to be the most updated version on the master branch.

If it helps, I also found that using the jspsych-survey-text would display the word "undefined" in the text box if I did not include `placeholder`as an explicit parameter. So I had to do them all like this:

```javascript
var state = {
 type: 'survey-text',
 questions: [{
   prompt:'In which state do you currently reside? Please write the full name of the state (e.g. New Mexico).',
   placeholder: ' ', // <-- Had to include this
   name: 'Q33'
 }]
};
```

So, there may be something going wrong with parameters that are supposed to default to `""` when undefined. This happens when running in both Chrome and Firefox.

Caitlin

**Comment by becky-gilbert** [comment]
I just tried to replicate this using jsPsych 6.1.0 and 6.0.5, on Chrome, Firefox and Edge. In all cases I got the automatic question naming, rather than 'undefined'. 

```javascript
[
	{
		"rt": 7097,
		"responses": "{\"Q0\":0,\"Q1\":0,\"Q2\":0,\"Q3\":0,\"Q4\":0}",
		"trial_type": "survey-likert",
		"trial_index": 0,
		"time_elapsed": 7111,
		"internal_node_id": "0.0-0.0"
	}
]
```
I wonder if you're right that this has something to do with using default values for unspecified parameters. This is something that the core jsPsych library now handles, but in older versions of jsPsych, it used to be that you had to manually set default values in the plugin files. So maybe one thing you could do is double check that your jspsych.js file matches the one from whatever relatively up-to-date release that you're using (either 6.1.0 or 6.0.5?) by doing a 'diff' in Git/GitHub or your text editor. Would you mind double checking that? Another thing you could try is downloading the release again to make sure that the jspsych.js file is the current version.
